fvm/state/state_replay_cache.go

```diff
@@ -0,0 +1,87 @@
+package state
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// StateReplayCache caches value read from state together with all register touches triggered
+// when reading that value from state. These register touches will be re/played on the target
+// view in every GetOrRetrieveValue value call.
+type StateReplayCache[T any] struct {
+	cachedValue T
+	cachedState *State
+}
+
+func NewStateReplayCache[T any]() *StateReplayCache[T] {
+	return &StateReplayCache[T]{}
+}
+
+// GetOrRetrieveValue returns a state value. When value is not cached, `retriver` closure will be
+// called to retrieve the value and the state delta will be captured and cached too; when value is
+// cached the cached one will be returned, and the cached state delta will be replayed on `view`.
+func (stateReplayCache *StateReplayCache[T]) GetOrRetrieveValue(
+	view View,
+	retriever func(*TransactionState) (T, error),
+	stateParams StateParameters,
+) (T, error) {
+	if view == nil {
+		return stateReplayCache.cachedValue, fmt.Errorf("view parameter should not be nil")
+	}
+	if retriever == nil {
+		return stateReplayCache.cachedValue, fmt.Errorf("retriever closure should not be nil")
+	}
+
+	transactionState := NewTransactionState(view, stateParams)
+
+	if stateReplayCache.cachedState == nil {
+		nestedTxId, err := transactionState.BeginNestedTransaction()
+		if err != nil {
+			return stateReplayCache.cachedValue, fmt.Errorf("failed to start a nested tx: %w", err)
+		}
+
+		// retrieve state value by calling retriever closure
+		stateReplayCache.cachedValue, err = retriever(transactionState)
```

Comment:
should we cache the error in stateReplayCache as well?

Reply:
here i return any error to caller to let them handle it immediately, so that in this cache we can assume that the cached state is not corrupted. it is mostly the question when\how we should handle errors from this cache. if we need to address a previously occurred error we need to cache that error as well.